test/graph_input_tests.cpp

```diff
@@ -0,0 +1,649 @@
+#include <catch2/catch_test_macros.hpp>
+
+#include "test_graph.h"
+
+#include <gdsb/graph_input.h>
+
+#include <sstream>
+#include <string>
+
+using namespace gdsb;
+
+TEST_CASE("read_ulong")
+{
+    unsigned long const a = 40;
+    unsigned long const b = 200;
+    unsigned long const c = 11;
+    unsigned long const d = 29848572984;
+
+    std::stringstream ss;
+    ss << a << " " << b << " " << c << " " << d << "\n";
+    std::string const test_string = ss.str();
+
+    char const* string_source = test_string.c_str();
+    char* string_position = nullptr;
+
+    unsigned long const a_read = read_ulong(string_source, &string_position);
+    string_source = string_position;
+    unsigned long const b_read = read_ulong(string_source, &string_position);
+    string_source = string_position;
+    unsigned long const c_read = read_ulong(string_source, &string_position);
+    string_source = string_position;
+    unsigned long const d_read = read_ulong(string_source, &string_position);
+
+    CHECK(a == a_read);
+    CHECK(b == b_read);
+    CHECK(c == c_read);
+    CHECK(d == d_read);
+}
+
+TEST_CASE("read_graph, edge_list")
+{
+    Edges32 edges;
+    auto emplace = [&](Vertex32 u, Vertex32 v) { edges.push_back(Edge32{ u, v }); };
+
+    WeightedEdges32 weighted_edges;
+    auto emplace_weighted = [&](Vertex32 u, Vertex32 v, Weight w) {
+        weighted_edges.push_back(WeightedEdge32{ u, Target32{ v, w } });
+    };
+
+    std::ifstream undirected_unweighted_temporal(graph_path + undirected_unweighted_temporal_reptilia_tortoise);
+    std::ifstream graph_input_weighted_temporal(graph_path + small_weighted_temporal_graph);
+    std::ifstream graph_input_unweighted_directed(graph_path + unweighted_directed_graph_enzymes);
+
+    SECTION("undirected, unweighted")
+    {
+        auto const [vertex_count, edge_count] =
+            read_graph<Vertex32, decltype(emplace), EdgeListUndirectedUnweightedStatic>(undirected_unweighted_temporal,
+                                                                                        std::move(emplace));
+
+        // directed: thus original edge count 103
+        CHECK(104 * 2 == edges.size());
+        CHECK(104 * 2 == edge_count);
+
+        bool edge_16_to_17_exists = [&]()
```

Comment:
There is no specific reason for choosing this edge except that it's not the first nor the last edge within the edge set that was added to the graph. Do you think choosing any random edge would be better? The issue with randomly choosing an edge is that it's less a unit test but more a robustness test. For unit tests I try to make any issue reproduce able.

Reply:
I see, for reproducibility one could use seeds combined with picking random edges, but as you alluded to, this might surpass the requirements for the unit-test.